runtime/gc_vlhgc/CopyForwardCompactGroup.hpp

```diff
@@ -126,12 +132,141 @@ class MM_CopyForwardCompactGroup
+	 */
+	MMINLINE bool abandonTLHRemainder(MM_EnvironmentVLHGC* env, uintptr_t age = 0)
+	{
+		bool abandoned = true;
+		if (NULL != _TLHRemainderBase) {
+			Assert_MM_true(NULL != _TLHRemainderTop);
+			/* make it a walkable hole */
+			env->_cycleState->_activeSubSpace->abandonHeapChunk(_TLHRemainderBase, _TLHRemainderTop);
+			MM_GCExtensions *extensions =MM_GCExtensions::getExtensions(env);
+			MM_HeapRegionDescriptorVLHGC *region = (MM_HeapRegionDescriptorVLHGC*)extensions->heapRegionManager->tableDescriptorForAddress(_TLHRemainderBase);
+
+			/* only preserve for non Nursery regions */
+			if ((age < extensions->tarokNurseryMaxAge._valueSpecified) || !region->containsObjects()) {
+				/* discard the remainder */
+				discardHeapChunk(_TLHRemainderBase, _TLHRemainderTop, region);
+				resetTLHRemainder();
+
+			} else {
+				abandoned = false;
+			}
+		} else {
+			Assert_MM_true(NULL == _TLHRemainderTop);
+		}
+		return abandoned;
+	}
+
+	MMINLINE void discardHeapChunk(void *base, void *top, MM_HeapRegionDescriptorVLHGC *region)
+	{
+		uintptr_t discardSize = ((uintptr_t)top - (uintptr_t)base);
+		_discardedBytes += discardSize;
+		region->getMemoryPool()->incrementDarkMatterBytes(discardSize);
+	}
+
+	/**
+	 * check and prepare TLHRemainder ready for using by collector
+	 *  1, reset TLHRemainder if it is invalid.
+	 *  2, reset TLHRemainder if it is in collection set
+	 *  3, align TLHRemainder with Card
+	 *  4, reset TLHRemainder if compactGroup doesn't match regionCompactGroup (the case: region overflow)
+     * @return region, if remainders are non-null, and regions is not already converted to survivor
+     * @return NULL, otherwise
+	 */
+	MMINLINE MM_HeapRegionDescriptorVLHGC* prepareTLHRemainder(MM_EnvironmentVLHGC* env, uintptr_t compactGroup)
+	{
+		MM_HeapRegionDescriptorVLHGC *region = NULL;
+		if (NULL != _TLHRemainderBase) {
+			bool needResetTLHRemainder = false;
+			MM_GCExtensions* extensions =MM_GCExtensions::getExtensions(env);
+			region = (MM_HeapRegionDescriptorVLHGC*)extensions->heapRegionManager->tableDescriptorForAddress(_TLHRemainderBase);
+			if (region->_markData._shouldMark || region->isFreeOrIdle() || (_gcIDForTLHRemainder <= region->_copyForwardData._lastGcIDForReclaim)) {
+				/* the region is part of collection set or has been reclaimed(swept/compacted/freed), reset TLHRemainder related with the region */
+				needResetTLHRemainder = true;
+			} else {
+				/* align TLHRemainder with card */
+				needResetTLHRemainder = !alignTLHRemainder(extensions->tlhSurvivorDiscardThreshold);
+				if (!needResetTLHRemainder) {
+					uintptr_t regionCompactGroup = MM_CompactGroupManager::getCompactGroupNumber(env, region);
+					if (regionCompactGroup != compactGroup) {
+						/* for region overflow case, region age has been set to tenure */
+						Assert_MM_true(extensions->tarokRegionMaxAge == region->getLogicalAge());
+						needResetTLHRemainder = true;
+					}
```

Comment:
is it valid to do ?

} else {
     assert (region age == compact group's age )
}